src/mlpack/methods/gmm/em_fit_impl.hpp

```diff
@@ -316,6 +325,63 @@ void EMFit<InitialClusteringType, CovarianceConstraintPolicy>::Serialize(
   ar & CreateNVP(constraint, "constraint");
 }
 
+template<typename InitialClusteringType, typename CovarianceConstraintPolicy>
+void EMFit<InitialClusteringType, CovarianceConstraintPolicy>::
+ArmadilloGMMWrapper(const arma::mat& observations,
+                    std::vector<distribution::GaussianDistribution>& dists,
+                    arma::vec& weights,
+                    const bool useInitialModel)
+{
+  arma::gmm_diag g;
+
+  // Warn the user that tolerance isn't used for convergence here.
+  Log::Warn << "GMM::Train(): tolerance ignored when training GMMs with "
+      << "DiagonalConstraint." << std::endl;
+
+  // If the initial clustering is the default k-means, we'll just use
+  // Armadillo's implementation.  If mlpack ever changes k-means defaults to use
+  // something that is reliably quicker than the Lloyd iteration k-means update,
+  // then this code maybe should be revisited.
+  if (!std::is_same<InitialClusteringType, mlpack::kmeans::KMeans<>>::value ||
+      useInitialModel)
+  {
+    // Use clusterer to get initial values.
+    if (!useInitialModel)
+      InitialClustering(observations, dists, weights);
+
+    // Assemble matrix of means.
+    arma::mat means(observations.n_rows, dists.size());
+    arma::mat covs(observations.n_rows, dists.size());
+    for (size_t i = 0; i < dists.size(); ++i)
+    {
+      means.col(i) = dists[i].Mean();
+      covs.col(i) = dists[i].Covariance().diag();
+    }
+
+    g.set_means(std::move(means));
+    g.set_dcovs(std::move(covs));
+    g.set_hefts(std::move(weights));
```

Comment:
You may want to use `g.set_params(means, dcovs, hefts)` here instead of setting them separately.  The `.set_means()` etc functions first check if the given parameters are consistent with the existing model (number of dimensions, gaussians, etc), so they may fail if the model isn't setup correctly in the first place.